src/common/os_ras_linux.c

```diff
@@ -0,0 +1,222 @@
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * os_ras_linux.c -- Linux ras abstraction layer
+ */
+#define _GNU_SOURCE
+
+#include <sys/types.h>
+#include <libgen.h>
+#include <limits.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "os.h"
+#include "os_ras.h"
+#include <ndctl/libndctl.h>
+#include <ndctl/libdaxctl.h>
+
+#define FOREACH_BUS_REGION_NAMESPACE(ctx, bus, region, ndns)	\
+	ndctl_bus_foreach(ctx, bus)				\
+		ndctl_region_foreach(bus, region)		\
+			ndctl_namespace_foreach(region, ndns)	\
+
+/*
+ * os_get_ndctl_set -- (internal) returns set of dimms where the pool is located
+ */
+static struct ndctl_interleave_set *
+os_get_ndctl_set(struct ndctl_ctx *ctx, os_stat_t st)
+{
+	dev_t dev;
+
+	if (S_ISCHR(st.st_mode)) {
+		dev = st.st_rdev;
+	} else {
+		dev = st.st_dev;
+	}
+
+	struct ndctl_bus *bus;
+	struct ndctl_region *region;
+	struct ndctl_namespace *ndns;
+
+	FOREACH_BUS_REGION_NAMESPACE(ctx, bus, region, ndns) {
+		struct ndctl_btt *btt = ndctl_namespace_get_btt(ndns);
+		struct ndctl_dax *dax = ndctl_namespace_get_dax(ndns);
+		struct ndctl_pfn *pfn = ndctl_namespace_get_pfn(ndns);
+		const char *devname;
+
+		if (btt) {
+			devname = ndctl_btt_get_block_device(btt);
+		} else if (pfn) {
+			devname = ndctl_pfn_get_block_device(pfn);
+		} else if (dax) {
+			struct daxctl_region *dax_region;
+			dax_region = ndctl_dax_get_daxctl_region(dax);
+			if (dax_region) {
+				struct daxctl_dev *dev;
+				dev = daxctl_dev_get_first(dax_region);
+				devname = daxctl_dev_get_devname(dev);
+			} else {
+				return NULL;
+			}
+		} else {
+			devname = ndctl_namespace_get_block_device(ndns);
+		}
+
+		if (*devname == '\0')
+			continue;
+
+		char *path;
+		os_stat_t stat;
+		if (asprintf(&path, "/dev/%s", devname) == -1)
```

Comment:
avoid functions that internally allocate memory. I'm fairly certain that you can simplify this code by having a static buffer of MAX_PATH size.